packages/origin-ui-core/src/components/Table/TableMaterial.tsx

```diff
@@ -121,6 +121,57 @@ export function TableMaterial<T extends readonly ITableColumn[]>(props: IProps<T
         }
     }
 
+    function renderActions(
+        actionsArr,
+        allowed,
+        row: TTableRow<GetReadonlyArrayItemType<T>['id']>,
+        rowId: number,
+        id: string,
+        classes: Record<'root' | 'tableWrapper' | 'tableCellWrappingActions', string>
+    ) {
+        if (actionsArr && !actionsArr.length) {
+            return <TableCell key={id}></TableCell>;
+        }
+        if (actionsArr && actionsArr.length > 0) {
+            const arrayCheck = Array.isArray(actionsArr[0]);
+            if (!arrayCheck) {
+                return (
+                    <TableCell key={id} className={classes.tableCellWrappingActions}>
+                        <Actions
+                            actions={
+                                allowed
+                                    ? actionsArr.filter((action) =>
+                                          allowed[(row as any).source]?.includes(action.id)
+                                      )
+                                    : actionsArr
+                            }
+                            id={id}
+                        />
+                    </TableCell>
+                );
+            }
+            if (arrayCheck && actionsArr[rowId].length > 0) {
+                return (
+                    <TableCell key={id} className={classes.tableCellWrappingActions}>
+                        <Actions
+                            actions={
+                                allowed
+                                    ? actionsArr[rowId].filter((action) =>
+                                          allowed[(row as any).source]?.includes(action.id)
+                                      )
+                                    : actionsArr[rowId]
+                            }
+                            id={id}
+                        />
+                    </TableCell>
+                );
+            }
+            if (arrayCheck && actionsArr[rowId].length === 0) {
+                return <TableCell key={id}></TableCell>;
+            }
+        }
+    }
```

Comment:
This method is way too complex and ambiguous, took me like a minute to understand what it does.

Consider this simplifying the implementation and making it more readable.
e.g. Name the variable `is2DArray` instead of `arrayCheck`.

```
if (!actionsArr?.length) {
    return <TableCell key={id}></TableCell>;
}

const is2DArray = Array.isArray(actionsArr[0]);

let finalActionsList;

if (!is2DArray) {
    finalActionsList = actionsArr;
} else if (actionsArr[rowId].length > 0) {
    finalActionsList = actionsArr[rowId];
} else {
    return <TableCell key={id}></TableCell>;
}

return (
    <TableCell key={id} className={classes.tableCellWrappingActions}>
        <Actions
            actions={
                allowed
                    ? finalActionsList.filter((action) =>
                            allowed[(row as any).source]?.includes(action.id)
                        )
                    : finalActionsList
            }
            id={id}
        />
    </TableCell>
);
```